go/vt/sqlparser/ast.go

```diff
@@ -2379,6 +2380,12 @@ func (node *ComparisonExpr) replace(from, to Expr) bool {
 	return replaceExprs(from, to, &node.Left, &node.Right, &node.Escape)
 }
 
+// IsImpossible returns true if the comparison in the expression can never evaluate to true.
+// Note that this is not currently exhaustive to ALL impossible comparisons.
+func (node *ComparisonExpr) IsImpossible() bool {
+	return node.Operator == NotEqualStr && reflect.DeepEqual(node.Left, node.Right)
```

Comment:
We've had a policy that `DeepEqual` shouldn't be used in the query serving path. This is because its cost is unpredictable. However, I'll be open to arguments about why it makes sense in this particular case. Let me know what you think.